Rename the default themes to gitea-light, gitea-dark, gitea-auto

silverwind opened the pull request:
Part of https://github.com/go-gitea/gitea/issues/27097:

- `gitea` theme is renamed to `gitea-light`
- `arc-green` theme is renamed to `gitea-dark`
- `auto` theme is renamed to `gitea-auto`

I put both themes in separate CSS files, removing all colors from the base CSS. Existing users will be migrated to the new theme names. The dark theme recolor will follow in a separate PR.

## :warning: BREAKING :warning:

1. If there are existing custom themes with the names `gitea-light` or `gitea-dark`, rename them before this upgrade and update the `theme` column in the `user` table for each affected user.
2. The theme in `<html>` has moved from `class="theme-name"` to `data-theme="name"`, existing customizations that depend on should be updated.

silverwind commented:
It seems `user` is a reserved word in postgres (where it should be escaped as `"user"`) and mssql (where it should be escaped as `[user]`. Does xorm handle this kind of escaping correctly when using backticks? I've seen these backticks in use in a number of other migrations.

lunny commented:
> `user` is a reserved word in postgres (where it should be escaped as `"user"`) and mssql (where it should be escaped as `[user]`). Does xorm handle this kind of escaping correctly when using backticks? I've seen these backticks in use in a number of other migrations, so I assume xorm abstracts this escaping.

Yes, xorm will convert ``word`` according different databases.

delvh commented:
Hmm… I'd like to wait with this PR until 1.21.0 has been released.
Otherwise, backports mostly won't be possible anymore.

silverwind commented:
>  Otherwise, backports mostly won't be possible anymore.

Which backports? Only backports that have CSS variable changes will present merge conflicts that are easily resolved.

caesar commented:
Should the `auto` theme also be renamed to `gitea-auto`?

silverwind commented:
I guess we could, I like such consistency.

denyskon commented:
I think gitea-* is better to indicate that these are our official themes

silverwind commented:
Renamed to `gitea-auto`. Latest change also fixed and simplified the logic in `head_style`, now matching the same in `head` that sets the theme class.

silverwind commented:
Theme name is now calculated in shared sub-template `templates/shared/misc/theme_name.tmpl`, eliminating the possibility for these to ever go out of sync.

silverwind commented on docs/content/administration/customizing-gitea.zh-cn.md:
```diff
@@ -86,5 +86,6 @@ Gitea 引用 `custom` 目录中的自定义配置文件来覆盖配置、模板
-`app.ini` [ui](administration/config-cheat-sheet.md#ui-ui) 部分的 `DEFAULT_THEME` 的值来变更至一个可用的 Gitea 外观。
+内置主题是“gitea-light”、“gitea-dark”和“gitea-auto”（自动适应操作系统设置）。
+
+默认主题可以通过 `app.ini` 的 [ui](administration/config-cheat-sheet.md#ui-ui) 部分中的 `DEFAULT_THEME` 进行更改。
```
Machine translations, please review.

silverwind commented:
Switched to helper function instead of sub-template because of different escaping needs and because go template does not support post-processing on `template` function return values.

silverwind commented:
One more change added: The theme name was previously on HTML `class="theme-<name>"`, and has moved to `data-theme="<name>"`. This was unused since the introduction of `--is-dark-theme` CSS variable, but I still find it good to retain under this new dedicated attribute for customization.

silverwind commented on modules/templates/helper.go:
```diff
@@ -131,7 +132,10 @@ func NewFuncMap() template.FuncMap {
+			if reflect.ValueOf(userTheme).Kind() == reflect.String && userTheme != "" {
+				return userTheme.(string)
+			}
 			return setting.UI.DefaultTheme
```
Maybe there are better ways than `reflect`. The function receives either nil or a string.

silverwind commented on templates/base/head_style.tmpl:
```diff
@@ -1,8 +1,2 @@
-{{else if ne DefaultTheme "gitea"}}
-	<link rel="stylesheet" href="{{AssetUrlPrefix}}/css/theme-{{DefaultTheme | PathEscape}}.css?v={{AssetVersion}}">
-{{end}}
+<link rel="stylesheet" href="{{AssetUrlPrefix}}/css/theme-{{ThemeName .SignedUser.Theme | PathEscape}}.css?v={{AssetVersion}}">
```
Yeah I can add that check just for safety. Previous code did not check it, so I assume it could have never been nil.

silverwind replied:
Refactored in https://github.com/go-gitea/gitea/pull/27419/commits/5ff07a9a04819b98415b969590e718fdb48c76c9.

silverwind commented on modules/templates/helper.go:
```diff
@@ -9,6 +9,7 @@ import (
 	"html"
 	"html/template"
 	"net/url"
+	"reflect"
```
```suggestion
```

silverwind commented on templates/status/500.tmpl:
```diff
@@ -1,12 +1,12 @@
 */}}
 <!DOCTYPE html>
-<html lang="{{.locale.Lang}}" class="theme-{{if .SignedUser.Theme}}{{.SignedUser.Theme}}{{else}}{{DefaultTheme}}{{end}}">
+<html lang="{{.locale.Lang}}" data-theme="{{ThemeName .SignedUser}}">
```
Not used at all by our first-party code, but it's useful to have imho, if only for customization. Previously the class was used at one point for Monaco, but that has been replaced by `--is-dark-theme`.
